Description: increases the coverage cap from 8.000 to 1.000.000

Index: samtools/bam2depth.c
===================================================================
--- samtools.orig/bam2depth.c
+++ samtools/bam2depth.c
@@ -106,6 +106,7 @@
 
 	// the core multi-pileup loop
 	mplp = bam_mplp_init(n, read_bam, (void**)data); // initialization
+	bam_mplp_set_maxcnt(mplp,1000000); // set maxdepth to 1M
 	n_plp = calloc(n, sizeof(int)); // n_plp[i] is the number of covering reads from the i-th BAM
 	plp = calloc(n, sizeof(void*)); // plp[i] points to the array of covering reads (internal in mplp)
 	while (bam_mplp_auto(mplp, &tid, &pos, n_plp, plp) > 0) { // come to the next covered position
